File: experiments/problem1_nonperiodicity/analyze_period_two_return_lift_profiles.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

constexpr std::array<char, 3> LETTERS{'t', 'p', 'u'};
constexpr std::array<int, 23> KAPPA_P{
    0, 1, 3, 7, 8, 8, 12, 13, 17, 17, 17, 21, 28, 30, 33, 34, 36, 40,
    40, 42, 47, 49, 51,
};
constexpr std::array<int, 23> KAPPA_U{
    0, 2, 2, 2, 7, 12, 14, 14, 14, 18, 19, 26, 27, 30, 30, 30, 30, 40,
    42, 42, 49, 52, 52,
};

uint64_t width_mask(int width) {
    if (width < 0 || width > 62) {
        throw std::invalid_argument("width must lie between zero and 62");
    }
    return width == 0 ? 0 : ((uint64_t{1} << width) - 1);
}

uint64_t forward_generator(char letter, uint64_t state, int width) {
    const uint64_t stepped =
        (state ^ ((state << 1) | (state << 2))) & width_mask(width);
    if (letter == 't') return stepped;
    if (letter == 'u') return stepped ^ 1;
    if (letter == 'p') {
        return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    }
    throw std::invalid_argument("unknown generator");
}

uint64_t inverse_t_mod(uint64_t output, int width) {
    output &= width_mask(width);
    uint64_t state = 0;
    for (int position = 0; position < width; ++position) {
        uint64_t previous = 0;
        if (position >= 1) previous |= (state >> (position - 1)) & 1;
        if (position >= 2) previous |= (state >> (position - 2)) & 1;
        const uint64_t bit = ((output >> position) & 1) ^ previous;
        state |= bit << position;
    }
    return state;
}

uint64_t inverse_generator_mod(char letter, uint64_t output, int width) {
    output &= width_mask(width);
    if (letter == 't') return inverse_t_mod(output, width);
    if (letter == 'u') return inverse_t_mod(output ^ 1, width);
    if (letter == 'p') {
        const uint64_t recovered_low_bit = (output & 1) ^ 1;
        const uint64_t adjusted =
            output ^ 1 ^ (recovered_low_bit == 0 ? 2 : 0);
        return inverse_t_mod(adjusted, width);
    }
    throw std::invalid_argument("unknown generator");
}

uint64_t fringe_step(uint64_t state) {
    const uint64_t packed = 1 + 2 * state;
    const uint64_t odd = packed ^ ((packed >> 1) | (packed >> 2));
    return (odd << 1) ^ (odd | (odd >> 1));
}

std::vector<char> actual_driver(int length) {
    uint64_t state = 0;
    std::vector<char> result;
    result.reserve(length);
    for (int index = 0; index < length; ++index) {
        result.push_back((state & 3) == 0 ? 'u' : 't');
        state = fringe_step(state);
    }
    return result;
}

uint64_t backward_zero_branch(char branch, uint64_t successor, int width) {
    const int inner_width = width - 2;
    uint64_t state = inverse_generator_mod(branch, successor, inner_width);
    state = inverse_generator_mod('p', state, inner_width);
    return ((state << 2) | 3) & width_mask(width);
}

uint64_t actual_survivor_residue(int depth) {
    const auto driver = actual_driver(depth);
    uint64_t state = 0;
    int precision = 0;
    for (auto iterator = driver.rbegin(); iterator != driver.rend(); ++iterator) {
        precision += 2;
        state = backward_zero_branch(*iterator, state, precision);
    }
    return state;
}

std::vector<std::pair<int, int>> actual_u_return_intervals(int count) {
    std::vector<int> positions;
    uint64_t state = 0;
    int position = 0;
    while (static_cast<int>(positions.size()) < count + 1) {
        if ((state & 3) == 0) positions.push_back(position);
        state = fringe_step(state);
        ++position;
    }
    std::vector<std::pair<int, int>> result;
    for (int index = 0; index < count; ++index) {
        const int gap = positions[index + 1] - positions[index];
        if (gap < 2 || gap > 5) {
            throw std::runtime_error("actual return gap escaped 2..5");
        }
        result.emplace_back(positions[index], positions[index + 1]);
    }
    return result;
}

uint64_t actual_block_code(int base_depth, int span) {
    const uint64_t lower = actual_survivor_residue(base_depth);
    const uint64_t upper = actual_survivor_residue(base_depth + span);
    if ((upper & width_mask(2 * base_depth)) != lower) {
        throw std::runtime_error("survivor residues failed to project");
    }
    return (upper - lower) >> (2 * base_depth);
}

uint64_t fnv1a64(const std::vector<uint8_t>& values) {
    uint64_t hash = 1469598103934665603ULL;
    for (uint8_t value : values) {
        hash ^= value;
        hash *= 1099511628211ULL;
    }
    return hash;
}

struct ProfileSummary {
    int minimum;
    int actual_length;
    int penalty;
    int minimizer_count;
    int maximum;
    uint64_t digest;
    std::map<int, int> histogram;
};

ProfileSummary return_lift_profile(
    int base_depth,
    int span,
    char phase,
    uint64_t actual_code
) {
    const int final_depth = base_depth + span;
    if (final_depth > 12) {
        throw std::invalid_argument("controlled final depth exceeds twelve");
    }
    const int width = 2 * final_depth;
    const uint32_t state_count = uint32_t{1} << width;
    const uint32_t block_count = uint32_t{1} << (2 * span);
    const uint64_t base = actual_survivor_residue(base_depth);

    std::vector<int32_t> target_code(state_count, -1);
    for (uint32_t code = 0; code < block_count; ++code) {
        const uint32_t target = static_cast<uint32_t>(
            base + (static_cast<uint64_t>(code) << (2 * base_depth))
        );
        target_code[target] = static_cast<int32_t>(code);
    }

    const uint32_t start = static_cast<uint32_t>(
        forward_generator(phase, 0, width)
    );
    std::vector<uint8_t> distance(state_count, 255);
    std::vector<uint32_t> queue;
    queue.reserve(state_count);
    std::vector<uint8_t> profile(block_count, 255);
    distance[start] = 0;
    queue.push_back(start);
    uint32_t remaining = block_count;
    if (target_code[start] >= 0) {
        profile[target_code[start]] = 1;
        --remaining;
    }

    for (std::size_t head = 0; head < queue.size() && remaining; ++head) {
        const uint32_t state = queue[head];
        const uint8_t next_distance = distance[state] + 1;
        for (char letter : LETTERS) {
            const uint32_t image = static_cast<uint32_t>(
                forward_generator(letter, state, width)
            );
            if (distance[image] != 255) continue;
            distance[image] = next_distance;
            queue.push_back(image);
            const int32_t code = target_code[image];
            if (code >= 0 && profile[code] == 255) {
                profile[code] = next_distance + 1;
                --remaining;
            }
        }
    }
    if (remaining) throw std::runtime_error("unreached block-lift target");

    const auto [minimum_it, maximum_it] =
        std::minmax_element(profile.begin(), profile.end());
    ProfileSummary summary{};
    summary.minimum = *minimum_it;
    summary.maximum = *maximum_it;
    summary.actual_length = profile.at(actual_code);
    summary.penalty = summary.actual_length - summary.minimum;
    summary.minimizer_count = std::count(
        profile.begin(), profile.end(), static_cast<uint8_t>(summary.minimum)
    );
    summary.digest = fnv1a64(profile);
    for (uint8_t value : profile) ++summary.histogram[value];
    return summary;
}

void print_histogram(const std::map<int, int>& histogram) {
    bool first = true;
    std::cout << '{';
    for (const auto& [distance, count] : histogram) {
        if (!first) std::cout << ',';
        first = false;
        std::cout << distance << ':' << count;
    }
    std::cout << '}';
}

}  // namespace

int main(int argc, char** argv) {
    int maximum_returns = 3;
    if (argc >= 2) maximum_returns = std::stoi(argv[1]);
    if (maximum_returns < 1 || maximum_returns > 3) {
        throw std::invalid_argument("maximum returns must lie in 1..3");
    }

    const auto intervals = actual_u_return_intervals(maximum_returns);
    std::array<int, 2> cumulative{0, 0};
    std::cout << "status finite-exhaustive\n";
    std::cout << "maximum_returns " << maximum_returns << '\n';

    for (int index = 0; index < maximum_returns; ++index) {
        const int left = intervals[index].first;
        const int right = intervals[index].second;
        const int gap = right - left;
        const int base_depth = left + 1;
        const int final_depth = right + 1;
        const uint64_t code = actual_block_code(base_depth, gap);

        std::cout << "return " << index
                  << " driver=" << left << ".." << right
                  << " gap=" << gap
                  << " depth=" << base_depth << ".." << final_depth
                  << " code=" << code << " digits=";
        for (int digit = 0; digit < gap; ++digit) {
            if (digit) std::cout << ',';
            std::cout << ((code >> (2 * digit)) & 3);
        }
        std::cout << '\n';

        for (int phase_index = 0; phase_index < 2; ++phase_index) {
            const char phase = phase_index == 0 ? 'p' : 'u';
            const auto summary =
                return_lift_profile(base_depth, gap, phase, code);
            const auto& kappa = phase == 'p' ? KAPPA_P : KAPPA_U;
            if (summary.minimum != kappa[base_depth]
                || summary.actual_length != kappa[final_depth]
                || summary.penalty != kappa[final_depth] - kappa[base_depth]) {
                throw std::runtime_error("profile theorem check failed");
            }
            cumulative[phase_index] += summary.penalty;
            std::cout << "phase " << phase
                      << " minimum=" << summary.minimum
                      << " actual=" << summary.actual_length
                      << " penalty=" << summary.penalty
                      << " minimizers=" << summary.minimizer_count
                      << " maximum=" << summary.maximum
                      << " fnv1a64=" << std::hex << std::setw(16)
                      << std::setfill('0') << summary.digest << std::dec
                      << std::setfill(' ') << " histogram=";
            print_histogram(summary.histogram);
            std::cout << '\n';
        }
    }

    const int first_depth = intervals.front().first + 1;
    const int final_depth = intervals.back().second + 1;
    if (cumulative[0] != KAPPA_P[final_depth] - KAPPA_P[first_depth]
        || cumulative[1] != KAPPA_U[final_depth] - KAPPA_U[first_depth]) {
        throw std::runtime_error("return penalties failed to telescope");
    }
    std::cout << "cumulative p=" << cumulative[0]
              << " u=" << cumulative[1] << '\n';
    std::cout << "scope no_infinite_penalty_recurrence_theorem\n";
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_return_lift_profiles.py
#!/usr/bin/env python3
"""Analyze witness-complexity penalties across actual ``u`` return blocks.

At successive ``u`` positions n_j of the actual period-two fringe schedule, set
L_j=n_j+1.  For phase a and a base-4 block code c of length r_j=n_(j+1)-n_j,
this module computes the shortest phase-a witness reaching

    X_(L_j) + c * 4**L_j  (mod 4**L_(j+1)).

The minimum over c is kappa_a(L_j), while the actual block coordinate is
kappa_a(L_(j+1)).  Their difference is the exact return-block penalty.  These
penalties telescope along the return sequence.

The finite campaign does not prove that return penalties are positive
infinitely often and does not solve Rule 30 center nonperiodicity.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from array import array
from collections import Counter
from typing import Any

LETTERS = ("t", "p", "u")
PHASES = ("p", "u")
DEFAULT_MAXIMUM_RETURNS = 2
ABSOLUTE_MAXIMUM_RETURNS = 3
ABSOLUTE_MAXIMUM_FINAL_DEPTH = 12

KAPPA_P = (
    0, 1, 3, 7, 8, 8, 12, 13, 17, 17, 17, 21, 28, 30, 33, 34, 36, 40,
    40, 42, 47, 49, 51,
)
KAPPA_U = (
    0, 2, 2, 2, 7, 12, 14, 14, 14, 18, 19, 26, 27, 30, 30, 30, 30, 40,
    42, 42, 49, 52, 52,
)


class ReturnLiftLimitError(RuntimeError):
    """Raised before an explicitly capped exhaustive campaign is exceeded."""


def width_mask(width: int) -> int:
    if width < 0:
        raise ValueError("width must be nonnegative")
    return (1 << width) - 1 if width else 0


def forward_generator(name: str, state: int, width: int) -> int:
    if state < 0 or width < 0:
        raise ValueError("state and width must be nonnegative")
    stepped = (state ^ ((state << 1) | (state << 2))) & width_mask(width)
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def inverse_t_mod(output: int, width: int) -> int:
    output &= width_mask(width)
    state = 0
    for position in range(width):
        previous = 0
        if position >= 1:
            previous |= (state >> (position - 1)) & 1
        if position >= 2:
            previous |= (state >> (position - 2)) & 1
        bit = ((output >> position) & 1) ^ previous
        state |= bit << position
    return state


def inverse_generator_mod(name: str, output: int, width: int) -> int:
    output &= width_mask(width)
    if name == "t":
        return inverse_t_mod(output, width)
    if name == "u":
        return inverse_t_mod(output ^ 1, width)
    if name == "p":
        recovered_low_bit = (output & 1) ^ 1
        adjusted = output ^ 1 ^ (2 if recovered_low_bit == 0 else 0)
        return inverse_t_mod(adjusted, width)
    raise ValueError("unknown generator")


def fringe_step(state: int) -> int:
    packed = 1 + 2 * state
    odd = packed ^ ((packed >> 1) | (packed >> 2))
    return (odd << 1) ^ (odd | (odd >> 1))


def actual_driver(length: int) -> list[str]:
    if length < 0:
        raise ValueError("length must be nonnegative")
    state = 0
    result: list[str] = []
    for _ in range(length):
        result.append("u" if state & 3 == 0 else "t")
        state = fringe_step(state)
    return result


def backward_zero_branch(branch: str, successor: int, width: int) -> int:
    if branch not in ("t", "u"):
        raise ValueError("branch must be t or u")
    if width < 2:
        raise ValueError("width must be at least two")
    inner_width = width - 2
    state = inverse_generator_mod(branch, successor, inner_width)
    state = inverse_generator_mod("p", state, inner_width)
    return ((state << 2) | 3) & width_mask(width)


def actual_survivor_residue(depth: int) -> int:
    if depth < 0:
        raise ValueError("depth must be nonnegative")
    state = 0
    precision = 0
    for branch in reversed(actual_driver(depth)):
        precision += 2
        state = backward_zero_branch(branch, state, precision)
    return state


def actual_u_return_intervals(count: int) -> tuple[tuple[int, int], ...]:
    if count < 0:
        raise ValueError("count must be nonnegative")
    positions: list[int] = []
    state = 0
    position = 0
    while len(positions) < count + 1:
        if state & 3 == 0:
            positions.append(position)
        state = fringe_step(state)
        position += 1
    intervals = tuple(zip(positions, positions[1:]))
    for left, right in intervals:
        if right - left not in (2, 3, 4, 5):
            raise AssertionError("actual u return gap escaped the exact 2..5 bound")
    return intervals


def actual_block_code(base_depth: int, span: int) -> int:
    if base_depth < 0 or span <= 0:
        raise ValueError("base depth must be nonnegative and span positive")
    lower = actual_survivor_residue(base_depth)
    upper = actual_survivor_residue(base_depth + span)
    if upper & width_mask(2 * base_depth) != lower:
        raise AssertionError("survivor residues failed to project")
    return (upper - lower) >> (2 * base_depth)


def return_lift_profile(base_depth: int, span: int, phase: str) -> tuple[int, ...]:
    if phase not in PHASES:
        raise ValueError("phase must be p or u")
    final_depth = base_depth + span
    if final_depth > ABSOLUTE_MAXIMUM_FINAL_DEPTH:
        raise ReturnLiftLimitError("final profile depth exceeds controlled maximum")

    width = 2 * final_depth
    state_count = 1 << width
    block_count = 1 << (2 * span)
    base = actual_survivor_residue(base_depth)
    targets = tuple(base + (code << (2 * base_depth)) for code in range(block_count))
    target_lookup = {target: code for code, target in enumerate(targets)}

    start = forward_generator(phase, 0, width)
    distances = bytearray(b"\xff") * state_count
    queue = array("I", [start])
    distances[start] = 0
    answers: list[int | None] = [None] * block_count
    if start in target_lookup:
        answers[target_lookup[start]] = 1
    remaining = sum(value is None for value in answers)

    head = 0
    while head < len(queue) and remaining:
        state = queue[head]
        head += 1
        next_distance = distances[state] + 1
        for letter in LETTERS:
            image = forward_generator(letter, state, width)
            if distances[image] != 255:
                continue
            distances[image] = next_distance
            queue.append(image)
            code = target_lookup.get(image)
            if code is not None and answers[code] is None:
                answers[code] = next_distance + 1
                remaining -= 1

    if remaining:
        raise AssertionError("positive generator graph failed to reach a block lift")
    return tuple(int(value) for value in answers)  # type: ignore[arg-type]


def summarize_profile(profile: tuple[int, ...], actual_code: int) -> dict[str, Any]:
    minimum = min(profile)
    histogram = Counter(profile)
    return {
        "minimum": minimum,
        "actual_length": profile[actual_code],
        "penalty": profile[actual_code] - minimum,
        "minimizer_count": sum(value == minimum for value in profile),
        "maximum": max(profile),
        "profile_sha256": hashlib.sha256(bytes(profile)).hexdigest(),
        "histogram": {str(key): histogram[key] for key in sorted(histogram)},
    }


def run_campaign(maximum_returns: int = DEFAULT_MAXIMUM_RETURNS) -> dict[str, Any]:
    if not 1 <= maximum_returns <= ABSOLUTE_MAXIMUM_RETURNS:
        raise ReturnLiftLimitError("return count outside controlled range")

    rows: list[dict[str, Any]] = []
    intervals = actual_u_return_intervals(maximum_returns)
    previous_final_depth: int | None = None
    cumulative = {phase: 0 for phase in PHASES}

    for return_index, (left, right) in enumerate(intervals):
        gap = right - left
        base_depth = left + 1
        final_depth = right + 1
        if previous_final_depth is not None and base_depth != previous_final_depth:
            raise AssertionError("return depth intervals failed to concatenate")
        previous_final_depth = final_depth

        actual_code = actual_block_code(base_depth, gap)
        actual_digits = [
            (actual_code >> (2 * index)) & 3 for index in range(gap)
        ]
        phase_rows: dict[str, Any] = {}
        for phase in PHASES:
            profile = return_lift_profile(base_depth, gap, phase)
            summary = summarize_profile(profile, actual_code)
            current = (KAPPA_P if phase == "p" else KAPPA_U)[base_depth]
            following = (KAPPA_P if phase == "p" else KAPPA_U)[final_depth]
            if summary["minimum"] != current:
                raise AssertionError("profile projection did not recover base complexity")
            if summary["actual_length"] != following:
                raise AssertionError("actual return coordinate did not recover final complexity")
            if summary["penalty"] != following - current:
                raise AssertionError("return penalty failed to equal complexity increment")
            cumulative[phase] += summary["penalty"]
            phase_rows[phase] = summary

        rows.append(
            {
                "return_index": return_index,
                "driver_start": left,
                "driver_end": right,
                "gap": gap,
                "base_depth": base_depth,
                "final_depth": final_depth,
                "actual_block_code": actual_code,
                "actual_block_digits_low_to_high": actual_digits,
                "by_phase": phase_rows,
            }
        )

    first_base_depth = rows[0]["base_depth"]
    last_final_depth = rows[-1]["final_depth"]
    for phase, sequence in (("p", KAPPA_P), ("u", KAPPA_U)):
        if cumulative[phase] != sequence[last_final_depth] - sequence[first_base_depth]:
            raise AssertionError("return penalties failed to telescope")

    payload: dict[str, Any] = {
        "status": "finite-exhaustive",
        "maximum_returns": maximum_returns,
        "rows": rows,
        "theorem_checks": {
            "return_depths": "L_j=n_j+1",
            "profile_minimum": "min_c Gamma_a(j,c)=kappa_a(L_j)",
            "actual_coordinate": "Gamma_a(j,c_j)=kappa_a(L_(j+1))",
            "penalty": "delta_(a,j)=kappa_a(L_(j+1))-kappa_a(L_j)",
            "telescoping": "kappa_a(L_J)=kappa_a(L_0)+sum_(j<J) delta_(a,j)",
            "boundedness": "kappa_a bounded iff all sufficiently late return penalties vanish",
        },
        "cumulative_penalties": cumulative,
        "scope_warning": (
            "finite return profiles do not prove positive penalties recur infinitely often, "
            "do not exclude period two, and do not solve Rule 30 center nonperiodicity"
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-returns",
        type=int,
        default=DEFAULT_MAXIMUM_RETURNS,
    )
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_returns), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
